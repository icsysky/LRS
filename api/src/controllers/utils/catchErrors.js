import handleError from 'api/controllers/utils/handleError';

export default handler => async (req, res) => {
  try {
    await handler(req, res);
  } catch (err) {
    return handleError(res, err);
  }
};
