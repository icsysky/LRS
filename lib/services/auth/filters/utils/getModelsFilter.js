import includes from 'lodash/includes';
import { ALL, SITE_ADMIN } from 'lib/constants/scopes';
import getTokenTypeFromAuthInfo
  from 'lib/services/auth/authInfoSelectors/getTokenTypeFromAuthInfo';
import getScopesFromAuthInfo
  from 'lib/services/auth/authInfoSelectors/getScopesFromAuthInfo';
import getOrgFilter from 'lib/services/auth/filters/getOrgFilter';
import NoAccessError from 'lib/errors/NoAccessError';
import { reduce } from 'bluebird';

export const getSiteAdminFilter = () => ({ authInfo }) => {
  const scopes = getScopesFromAuthInfo(authInfo);
  if (includes(scopes, SITE_ADMIN)) {
    return getOrgFilter(authInfo);
  }
};

export const checkAllowedTokenType =
  ({ allowedTokenTypes = ['organisation', 'client'] }) => ({ authInfo }) => {
    const tokenType = getTokenTypeFromAuthInfo(authInfo);
    if (
      !includes(allowedTokenTypes, tokenType)
    ) {
      throw new NoAccessError();
    }
  };

export const checkAllScope = () => ({ authInfo }) => {
  const scopes = getScopesFromAuthInfo(authInfo);

  if (includes(scopes, ALL)) {
    return getOrgFilter(authInfo);
  }
};

export default scopeConstants => async (options) => {
  const { filters } = scopeConstants;

  return reduce(filters, async (result, item) => {
    if (result) {
      return result;
    }

    const out = await item(scopeConstants)(options);

    return out;
  }, null);
};
