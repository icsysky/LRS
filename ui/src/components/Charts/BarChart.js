import React, { Component, PropTypes } from 'react';
import { List } from 'immutable';
import { AutoSizer } from 'react-virtualized';
import { BarChart as Chart, XAxis, YAxis } from 'recharts';
import { compose } from 'recompose';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import NoData from 'ui/components/Graphs/NoData';
import { Button } from 'react-toolbox/lib/button';
import {
  getResultsData,
  getShortModel,
  getChartData,
  getDomain,
  hasData,
  renderTooltips,
  renderBars,
  renderLegend,
  hiddenSeriesState
} from './Chart';
import styles from './styles.css';

const enhance = compose(
  withStyles(styles),
  hiddenSeriesState
);

class BarChart extends Component {
  static propTypes = {
    results: PropTypes.instanceOf(List),
    labels: PropTypes.instanceOf(List),
    colors: PropTypes.instanceOf(List),
    stacked: PropTypes.bool,
    axesLabels: PropTypes.instanceOf(Object),
    chartWrapperFn: PropTypes.instanceOf(Function)
  }

  static defaultProps = {
    chartWrapperFn: component => (<AutoSizer>{component}</AutoSizer>)
  }

  state = {
    activePage: 0,
  }

  displayPrevPage = () => this.setState({ activePage: this.state.activePage - 1 })
  displayNextPage = () => this.setState({ activePage: this.state.activePage + 1 })
  getDataChunk = data => page => data.slice(10 * page, 10 * (page + 1))
  getPages = data => Math.ceil(data.size / 10)
  hasPrevPage = pages => page => pages > 0 && page > 0
  hasNextPage = pages => page => pages > 0 && page < pages - 1
  sortData = data => data.sortBy(e => -e.get('total'))
  getSortedData = results => labels => (
    this.sortData(getResultsData(results)(labels))
  )

  renderPrevButton = () => (
    <span className={styles.prevButton}>
      <Button
        raised
        label="Previous"
        onMouseUp={this.displayPrevPage}
        style={styles.button}
        icon={<i className="icon ion-chevron-left" />} />
    </span>
  )

  renderNextButton = () => (
    <span className={styles.nextButton}>
      <Button
        raised
        label="Next"
        onMouseUp={this.displayNextPage}
        style={styles.button}
        icon={<i className="icon ion-chevron-right" />} />
    </span>
  )

  renderBarChart = colors => labels => data => stacked => page => ({ width, height }) => (
    <Chart
      data={getChartData(this.getDataChunk(data)(page), this.props.hiddenSeries)}
      width={width}
      height={height}
      layout="vertical">
      <YAxis
        dataKey="cellId"
        tickFormatter={getShortModel(data)}
        type="category"
        width={90} />
      <XAxis type="number" domain={getDomain(data)} />
      {renderLegend(labels, this.props.toggleHiddenSeries)}
      {renderBars(colors)(labels)(stacked)}
      {renderTooltips(data, this.props.hiddenSeries)}
    </Chart>
    )

  renderResults = results => colors => labels => (stacked) => {
    const { activePage } = this.state;
    const data = this.getSortedData(results)(labels);
    const pages = this.getPages(data);

    return (
      <div className={`${styles.chart}`}>
        <div className={`${styles.buttons}`}>
          {this.hasPrevPage(pages)(activePage) && this.renderPrevButton()}
          {this.hasNextPage(pages)(activePage) && this.renderNextButton()}
        </div>
        <div className={`${styles.xAxisLabel} ${styles.withPrevNext} clearfix`}>
          <span className={styles.xAxis}>
            {this.props.axesLabels.xLabel || 'X Axis'}
          </span>
        </div>
        <div className={`${styles.barContainer}`}>
          <div className={styles.yAxisLabel}>
            <span className={styles.yAxis}>
              {this.props.axesLabels.yLabel || 'Y Axis'}
            </span>
          </div>
          <div className={styles.chartWrapper}>
            {this.props.chartWrapperFn((this.renderBarChart(colors)(labels)(data)(stacked)(activePage)))}
          </div>
        </div>
      </div>
    );
  }

  render = () => {
    const { results, labels, stacked, colors } = this.props;
    return (
      hasData(this.props.results)
      ? this.renderResults(results)(colors)(labels)(stacked)
      : <NoData />
    );
  }
}

export default enhance(BarChart);
