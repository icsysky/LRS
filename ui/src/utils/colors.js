const hexToRgb = (hex) => {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16)
  } : null;
};

export default ({
  hexColor = '#f5d76e',
  range,
  value,
  minOpacity = 0.2,
  maxOpacity = 1,
}) => {
  const rgb = hexToRgb(hexColor);
  const opacityRange = maxOpacity - minOpacity;
  const opacityInterval = (range > 1 ? opacityRange / (range - 1) : opacityRange);
  const opacity = (opacityInterval * value) + minOpacity;
  return `rgba(${rgb.r},${rgb.g},${rgb.b},${opacity})`;
};

