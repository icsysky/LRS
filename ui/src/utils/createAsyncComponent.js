import React, { Component } from 'react';
import Promise from 'bluebird';

Promise.config({ cancellation: true });

export default ({ loader, preview = null }) => class asyncComponent extends Component {
  constructor() {
    super();
    this.state = {
      preview
    };
  }

  componentWillMount = () => {
    this.blueBirdLoader = Promise.resolve(loader);
    this.blueBirdLoader.then((loadedComponent) => {
      this.setState({ component: loadedComponent.default || loadedComponent });
    });
  }

  componentWillUnmount() {
    this.blueBirdLoader.cancel();
  }

  render = () => {
    if (this.state.component) {
      return <this.state.component {...this.props} />;
    }
    return this.state.preview;
  }
};
