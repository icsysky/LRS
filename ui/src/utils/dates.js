import {
  LAST_30_DAYS,
  LAST_7_DAYS,
  LAST_2_MONTHS,
  LAST_6_MONTHS,
  LAST_24_HOURS,
  LAST_1_YEAR,
  LAST_2_YEARS
} from 'ui/utils/constants';
import moment from 'moment';

export const periodToDate = (datePeriod, today) => {
  const today2 = moment(today); // Clones the date so it's not mutated.
  switch (datePeriod) {
    case LAST_24_HOURS: return today2.subtract(24, 'hours');
    case LAST_7_DAYS: default: return today2.subtract(7, 'days');
    case LAST_30_DAYS: return today2.subtract(30, 'days');
    case LAST_2_MONTHS: return today2.subtract(2, 'months');
    case LAST_6_MONTHS: return today2.subtract(6, 'months');
    case LAST_1_YEAR: return today2.subtract(1, 'years');
    case LAST_2_YEARS: return today2.subtract(2, 'years');
  }
};

export default (...args) => ({ periodToDate: periodToDate(...args).toISOString() });
