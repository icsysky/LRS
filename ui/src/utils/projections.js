/* eslint-disable import/prefer-default-export */
export const sections = {
  source: [
    '$statement',
    '$statement.actor',
    '$statement.actor.account',
    '$statement.actor.account.name',
    '$statement.actor.account.homePage',
    '$statement.actor.mbox',
    '$statement.actor.mbox_sha1sum',
    '$statement.actor.member',
    '$statement.actor.name',
    '$statement.actor.objectType',
    '$statement.actor.openid',
    '$statement.authority',
    '$statement.authority.mbox',
    '$statement.authority.name',
    '$statement.authority.objectType',
    '$statement.context',
    '$statement.context.contextActivities',
    '$statement.context.contextActivities.category',
    '$statement.context.contextActivities.grouping',
    '$statement.context.contextActivities.parent',
    '$statement.context.contextActivities.other',
    '$statement.context.instructor',
    '$statement.context.language',
    '$statement.context.platform',
    '$statement.context.registration',
    '$statement.context.statement',
    '$statement.context.statement.id',
    '$statement.context.statement.objectType',
    '$statement.context.team',
    '$statement.id',
    '$statement.object',
    '$statement.object.id',
    '$statement.object.definition',
    '$statement.object.definition.choices',
    '$statement.object.definition.correctResponsesPattern',
    '$statement.object.definition.description',
    '$statement.object.definition.extensions',
    '$statement.object.definition.interactionType',
    '$statement.object.definition.moreInfo',
    '$statement.object.definition.name',
    '$statement.object.definition.scale',
    '$statement.object.definition.source',
    '$statement.object.definition.target',
    '$statement.object.definition.type',
    '$statement.object.objectType',
    '$statement.result',
    '$statement.result.completion',
    '$statement.result.duration',
    '$statement.result.extensions',
    '$statement.result.response',
    '$statement.result.success',
    '$statement.stored',
    '$statement.timestamp',
    '$statement.verb',
    '$statement.verb.id',
    '$statement.verb.display',
    '$statement.version'
  ],
  STATEMENTS: [
    '$_id',
    '$count'
  ],
  LEADERBOARD: [
    '$_id',
    '$_id.account',
    '$_id.account.name',
    '$_id.account.homePage',
    '$_id.mbox',
    '$_id.mbox_sha1sum',
    '$_id.member',
    '$_id.name',
    '$_id.objectType',
    '$_id.openid',
    '$count'
  ],
  POPULARACTIVITIES: [
    '$activity',
    '$activity.id',
    '$activity.definition',
    '$activity.definition.choices',
    '$activity.definition.correctResponsesPattern',
    '$activity.definition.description',
    '$activity.definition.extensions',
    '$activity.definition.interactionType',
    '$activity.definition.moreInfo',
    '$activity.definition.name',
    '$activity.definition.scale',
    '$activity.definition.source',
    '$activity.definition.target',
    '$activity.definition.type',
    '$activity.objectType',
    '$count'
  ],
  XVSY: [
    '$_id',
    '$_id.account',
    '$_id.account.name',
    '$_id.account.homePage',
    '$_id.mbox',
    '$_id.mbox_sha1sum',
    '$_id.member',
    '$_id.name',
    '$_id.objectType',
    '$_id.openid',
    '$count'
  ],
  FREQUENCY: [
    '$_id',
    '$count'
  ],
};
