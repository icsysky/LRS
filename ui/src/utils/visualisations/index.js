import pipelinesFromQueries from './pipelinesFromQueries';
import localOptions from './localOptions';

export {
  pipelinesFromQueries,
  localOptions
};
