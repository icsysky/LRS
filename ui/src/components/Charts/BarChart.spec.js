import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { withInsertCSS } from 'ui/utils/hocs';
import { List, fromJS } from 'immutable';
import BarChart from './BarChart';

const WrappedBarChart = withInsertCSS(BarChart);

test('BarChartResults should render', () => {
  const barChart = ReactTestRenderer.create(
    <WrappedBarChart
      results={new List()} />
  ).toJSON();

  expect(barChart).toMatchSnapshot();
});

test('BarChartResult should render axes', () => {
  const mockResult = fromJS([[{
    '1970-01-01': {
      _id: '1970-01-01',
      count: 1,
      model: '1970-01-01'
    }
  }]]);

  const barChart = ReactTestRenderer.create(
    <WrappedBarChart
      results={mockResult}
      labels={fromJS([undefined])}
      axesLabels={{
        yLabel: 'THE-Y',
        xLabel: 'THE-X'
      }}
      chartWrapperFn={() => (<div />)} />
  ).toJSON();

  expect(barChart).toMatchSnapshot();
});
