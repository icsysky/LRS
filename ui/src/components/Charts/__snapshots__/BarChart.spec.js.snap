// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChartResult should render axes 1`] = `
<div
  className="chart"
>
  <div
    className="buttons"
  />
  <div
    className="xAxisLabel withPrevNext clearfix"
  >
    <span
      className="xAxis"
    >
      THE-X
    </span>
  </div>
  <div
    className="barContainer"
  >
    <div
      className="yAxisLabel"
    >
      <span
        className="yAxis"
      >
        THE-Y
      </span>
    </div>
    <div
      className="chartWrapper"
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`BarChartResults should render 1`] = `
<h4
  style={
    Object {
      "marginTop": 20,
      "textAlign": "center",
    }
  }
>
  No data in this time range.
</h4>
`;
